File: doc/integrations/drp/DRP-go/drpmesh/endpoint.go
package drpmesh

import (
	"encoding/json"
	"fmt"

	"github.com/gorilla/websocket"
)

// ConnectionStats provides latency and uptime stats
type ConnectionStats struct {
	pingTimeMs    int
	uptimeSeconds int
}

// EndpointMethod defines the interface for a DRP Endpoint method
type EndpointMethod func(*CmdParams, EndpointInterface, *int) interface{}

// EndpointAuthInfo tracks the auth info provided by a remote Node
type EndpointAuthInfo struct {
	Type     string
	Value    string
	UserInfo interface{}
}

// EndpointInterface declares the set of functions that should be implemented for any Endpoint object
type EndpointInterface interface {
	GetID() *string
	GetType() string
	GetToken() int
	AddReplyHandler() int
	DeleteReplyHandler(int)
	RegisterMethod(string, EndpointMethod)
	SendPacketBytes([]byte)
	SendCmd(string, string, interface{}, *int, *RouteOptions, *string)
	SendCmdAwait(string, string, interface{}, *RouteOptions, *string) *ReplyIn
	IsServer() bool
	ConnectionStats() ConnectionStats
	GetEndpointCmds() map[string]EndpointMethod
	IsReady() bool
	IsConnecting() bool
}

// Endpoint - DRP endpoint
type Endpoint struct {
	wsConn            *websocket.Conn
	drpNode           *Node
	EndpointID        *string
	EndpointType      string
	EndpointCmds      map[string]EndpointMethod
	AuthInfo          EndpointAuthInfo
	ReplyHandlerQueue map[int](chan *ReplyIn)
	TokenNum          int
	Subscriptions     interface{}
	openCallback      *func()
	closeCallback     *func()
	closeChan         chan bool

	//sendChan chan interface{}
}

// Init initializes Endpoint attributes
func (e *Endpoint) Init() {
	e.EndpointCmds = make(map[string]EndpointMethod)
	e.ReplyHandlerQueue = make(map[int](chan *ReplyIn))
	e.closeChan = make(chan bool)
	//e.sendChan = make(chan interface{}, 100)
	e.TokenNum = 1
}

// GetID returns the ID of the Endpoint
func (e *Endpoint) GetID() *string {
	return e.EndpointID
}

// GetType returns the type of the Endpoint
func (e *Endpoint) GetType() string {
	return e.EndpointType
}

// GetToken returns the next token to be used for the Endpoint
func (e *Endpoint) GetToken() int {
	//returnToken := strconv.Itoa(e.TokenNum)
	returnToken := e.TokenNum
	e.TokenNum++
	return returnToken
}

// AddReplyHandler is used to track responses for a command
func (e *Endpoint) AddReplyHandler() int {
	replyToken := e.GetToken()
	e.ReplyHandlerQueue[replyToken] = make(chan *ReplyIn)
	return replyToken
}

// DeleteReplyHandler removes a reply handler
func (e *Endpoint) DeleteReplyHandler(handlerToken int) {
	delete(e.ReplyHandlerQueue, handlerToken)
	return
}

// RegisterMethod adds a command that is allowed to be executed by the remote Endpoint
func (e *Endpoint) RegisterMethod(methodName string, method EndpointMethod) {
	e.EndpointCmds[methodName] = method
}

// SendPacketBytes abstracts sending functions from the communication channel functions
func (e *Endpoint) SendPacketBytes(drpPacketBytes []byte) {
	// This function listend to e.sendChan and sends commands
	wsSendErr := e.wsConn.WriteMessage(websocket.TextMessage, drpPacketBytes)
	if wsSendErr != nil {
		e.drpNode.Log(fmt.Sprint("error writing message to WS channel:", wsSendErr), false)
		//return wsSendErr
	}
}

// SendCmd sends a command to a remote Endpoint
func (e *Endpoint) SendCmd(serviceName string, methodName string, cmdParams interface{}, token *int, routeOptions *RouteOptions, serviceInstanceID *string) {
	sendCmd := &CmdOut{}
	sendCmd.ServiceName = &serviceName
	sendCmd.Type = "cmd"
	sendCmd.Method = &methodName
	sendCmd.Params = cmdParams
	sendCmd.Token = token
	sendCmd.RouteOptions = routeOptions
	sendCmd.ServiceInstanceID = serviceInstanceID

	packetBytes := sendCmd.ToJSON()
	e.SendPacketBytes(packetBytes)
	//e.sendChan <- *sendCmd
}

// SendCmdAwait sends a command to a remote Endpoint and awaits a response
func (e *Endpoint) SendCmdAwait(serviceName string, cmdName string, cmdParams interface{}, routeOptions *RouteOptions, serviceInstanceID *string) *ReplyIn {
	replyToken := e.AddReplyHandler()
	e.SendCmd(serviceName, cmdName, cmdParams, &replyToken, routeOptions, serviceInstanceID)
	responseData := <-e.ReplyHandlerQueue[replyToken]

	return responseData
}

// SendReply returns data to a remote Endpoint which originally executed a command
func (e *Endpoint) SendReply(replyToken *int, returnStatus int, returnPayload interface{}) {
	replyCmd := &ReplyOut{}
	replyCmd.Type = "reply"
	replyCmd.Token = replyToken
	replyCmd.Status = returnStatus
	replyCmd.Payload = returnPayload

	packetBytes := replyCmd.ToJSON()
	e.drpNode.Log(fmt.Sprintf("SendReply -> %s", string(packetBytes)), true)
	e.SendPacketBytes(packetBytes)
	//e.sendChan <- *replyCmd
}

// ProcessCmd processes an inbound packet as a Cmd
func (e *Endpoint) ProcessCmd(msgIn *Cmd) {
	cmdResults := make(map[string]interface{})
	cmdResults["status"] = 0
	cmdResults["output"] = nil
	if _, ok := e.EndpointCmds[*msgIn.Method]; ok {
		// Execute command
		e.drpNode.Log(fmt.Sprintf("Executing method '%s'...", *msgIn.Method), true)
		cmdResults["output"] = e.EndpointCmds[*msgIn.Method](msgIn.Params, e, msgIn.Token)
		cmdResults["status"] = 1
	} else {
		cmdResults["output"] = "Endpoint does not have this method"
		e.drpNode.Log(fmt.Sprintf("Remote Endpoint tried to execute unrecognized method '%s'", *msgIn.Method), true)
	}
	e.SendReply(msgIn.Token, cmdResults["status"].(int), cmdResults["output"])
}

// ProcessReply processes an inbound packet as a Reply
func (e *Endpoint) ProcessReply(msgIn *ReplyIn) {
	e.ReplyHandlerQueue[*msgIn.Token] <- msgIn

	// If the receive is complete, delete handler
	if msgIn.Status < 2 {
		e.DeleteReplyHandler(*msgIn.Token)
	}
}

// ShouldRelay determines whether or not an inbound packet should be relayed
func (e *Endpoint) ShouldRelay(msgIn *PacketIn) bool {
	var shouldForward = false
	if msgIn.RouteOptions != nil && msgIn.RouteOptions.TgtNodeID != nil && *msgIn.RouteOptions.TgtNodeID != e.drpNode.NodeID {
		shouldForward = true
	}
	return shouldForward
}

// ReceiveMessage determines whether a packet should be relayed or processed locally
func (e *Endpoint) ReceiveMessage(rawMessage []byte) {

	packetIn := &PacketIn{}
	err := json.Unmarshal(rawMessage, packetIn)
	if err != nil {
		e.drpNode.Log(fmt.Sprintf("ReceiveMessage Packet unmarshal error: %s", err), false)
		return
	}

	if e.ShouldRelay(packetIn) {
		e.RelayPacket(packetIn)
		return
	}

	switch packetIn.Type {
	case "cmd":
		cmdPacket := &Cmd{}
		err := json.Unmarshal(rawMessage, cmdPacket)
		if err != nil {
			e.drpNode.Log(fmt.Sprintf("ReceiveMessage Cmd unmarshal error: %s", err), false)
			return
		}
		e.ProcessCmd(cmdPacket)
	case "reply":
		replyPacket := &ReplyIn{}
		err := json.Unmarshal(rawMessage, replyPacket)
		if err != nil {
			e.drpNode.Log(fmt.Sprintf("ReceiveMessage Reply unmarshal error: %s", err), false)
			return
		}
		e.ProcessReply(replyPacket)
	}
}

// RelayPacket routes a packet to another Node
func (e *Endpoint) RelayPacket(packetIn *PacketIn) {
	thisEndpoint := e
	var errMsg *string = nil

	// Validate sending endpoint
	if thisEndpoint.EndpointID == nil {

		// Sending endpoint has not authenticated
		tmpErr := "sending endpoint has not authenticated"
		errMsg = &tmpErr

		// Validate source node
	} else if !thisEndpoint.drpNode.TopologyTracker.ValidateNodeID(*packetIn.RouteOptions.SrcNodeID) {

		// Source NodeID is invalid
		tmpErr := fmt.Sprintf("srcNodeID %s not found", *packetIn.RouteOptions.SrcNodeID)
		errMsg = &tmpErr

		// Validate destination node
	} else if !thisEndpoint.drpNode.TopologyTracker.ValidateNodeID(*packetIn.RouteOptions.TgtNodeID) {
		// Target NodeID is invalid
		tmpErr := fmt.Sprintf("tgtNodeID %s not found", *packetIn.RouteOptions.TgtNodeID)
		errMsg = &tmpErr
	}

	if errMsg != nil {
		thisEndpoint.drpNode.Log(fmt.Sprintf("Could not relay message: %s", *errMsg), false)
		return
	}

	nextHopNodeID := thisEndpoint.drpNode.TopologyTracker.GetNextHop(*packetIn.RouteOptions.TgtNodeID)
	targetNodeEndpoint := thisEndpoint.drpNode.VerifyNodeConnection(*nextHopNodeID)

	// Add this node to the routing history
	packetIn.RouteOptions.RouteHistory = append(packetIn.RouteOptions.RouteHistory, thisEndpoint.drpNode.NodeID)

	// Repackage
	packetBytesOut := []byte{}
	switch packetIn.Type {
	case "cmd":
		cmdPacket := &Cmd{
			BasePacket{
				packetIn.Type,
				packetIn.RouteOptions,
				packetIn.Token,
			},
			packetIn.Method,
			packetIn.Params,
			packetIn.ServiceName,
			packetIn.ServiceInstanceID,
		}
		packetBytesOut = cmdPacket.ToJSON()
	case "reply":
		replyPacket := &ReplyIn{
			BasePacket{
				packetIn.Type,
				packetIn.RouteOptions,
				packetIn.Token,
			},
			packetIn.Status,
			packetIn.Payload,
		}
		packetBytesOut = replyPacket.ToJSON()
	}

	// Send packet to next hop
	targetNodeEndpoint.SendPacketBytes(packetBytesOut)

	// Increment local Node's PacketRelayCount
	thisEndpoint.drpNode.PacketRelayCount++

	return
}

// GetCmds returns the list of method available for the peer Endpoint to execute
func (e *Endpoint) GetCmds() interface{} {
	keys := make([]string, 0)
	for key := range e.EndpointCmds {
		keys = append(keys, key)
	}
	return keys
}

// OpenHandler specifies actions to be taken after a connection is opened
func (e *Endpoint) OpenHandler() {
}

// CloseHandler specifies actions to be taken after a connection is closed
func (e *Endpoint) CloseHandler(code int, text string) error {
	e.closeChan <- true
	return nil
}

// ErrorHandler specifies actions to be taken after a connection encounters an error
func (e *Endpoint) ErrorHandler() {
}

// StartListening begins listening loop on wsConn
func (e *Endpoint) StartListening(wsConn *websocket.Conn) {
	e.wsConn = wsConn
	e.wsConn.SetCloseHandler(e.CloseHandler)

	// Start receive loop
	go func() {
		for {
			_, p, err := e.wsConn.ReadMessage()
			if err != nil {
				e.drpNode.Log(fmt.Sprintf("Could not read from wsConn: %s", err), true)
				break
			} else {
				// Need to update this so the receive loop isn't tied up processing inbound commands
				e.ReceiveMessage(p)
			}
		}
	}()
}

// IsServer tells whether or not this endpoint is the server side of the connection
func (e *Endpoint) IsServer() bool {
	return false
}

// PingTime TO DO - IMPLEMENT
func (e *Endpoint) PingTime() int {
	return 0
}

// UpTime TO DO - IMPLEMENT
func (e *Endpoint) UpTime() int {
	return 0
}

// IsReady tells whether or not the Endpoint's socket connection is ready to communicate
// TO DO - IMPLEMENT
func (e *Endpoint) IsReady() bool {
	return true
}

// IsConnecting tells whether or not the Endpoint's socket connection is attempting to establish a connection
// TO DO - IMPLEMENT
func (e *Endpoint) IsConnecting() bool {
	return false
}

// ConnectionStats returns uptime and latency info
func (e *Endpoint) ConnectionStats() ConnectionStats {
	return ConnectionStats{
		e.PingTime(),
		e.UpTime(),
	}
}

// GetEndpointCmds returns the commands available to be executed on this endpoint
func (e *Endpoint) GetEndpointCmds() map[string]EndpointMethod {
	return e.EndpointCmds
}

File: doc/integrations/drp/DRP-nodejs/drp-mesh/lib/endpoint.js
'use strict';

// Had to remove this so we don't have a circular eval problem
//const DRP_Node = require('./node');
//const DRP_SubscribableSource = require('./subscription').DRP_SubscribableSource;
const DRP_Subscriber = require('./subscription').DRP_Subscriber;
const { DRP_Packet, DRP_Cmd, DRP_Reply, DRP_RouteOptions } = require('./packet');

const WebSocket = require('ws');

class DRP_Endpoint {
    /**
     * 
     * @param {Websocket} wsConn Websocket connection
     * @param {DRP_Node} drpNode DRP Node
     * @param {string} endpointID Remote Endpoint ID
     * @param {string} endpointType Remote Endpoint Type
     */
    constructor(wsConn, drpNode, endpointID, endpointType) {
        let thisEndpoint = this;
        /** @type {WebSocket} */
        this.wsConn = wsConn || null;
        /** @type {DRP_Node} */
        this.DRPNode = drpNode;
        if (this.wsConn) {
            this.wsConn.drpEndpoint = this;
        }
        this.EndpointID = endpointID || null;
        this.EndpointType = endpointType || null;
        this.EndpointCmds = {};

        this.AuthInfo = {
            type: null,
            value: null,
            userInfo: null
        };

        /** @type Object<number,function> */
        this.ReplyHandlerQueue = {};

        this.TokenNum = 1;

        /** @type {Object.<string,DRP_Subscriber>} */
        this.Subscriptions = {};
        /** @type {function} */
        this.openCallback;
        /** @type {function} */
        this.closeCallback;
        this.RegisterMethod("getCmds", "GetCmds");

        this.RemoteAddress = this.RemoteAddress;
        this.RemotePort = this.RemotePort;
        this.RemoteFamily = this.RemoteFamily;
    }

    GetToken() {
        let token = this.TokenNum;
        this.TokenNum++;
        return token;
    }

    AddReplyHandler(callback) {
        let token = this.GetToken();
        this.ReplyHandlerQueue[token] = callback;
        return token;
    }

    DeleteReplyHandler(token) {
        delete this.ReplyHandlerQueue[token];
    }

    /**
     * Register Endpoint Command
     * @param {string} methodName Method Name
     * @param {function(Object.<string,object>, DRP_Endpoint, string)} method Function
     */
    RegisterMethod(methodName, method) {
        let thisEndpoint = this;
        // Need to do sanity checks; is the method actually a method?
        if (typeof method === 'function') {
            thisEndpoint.EndpointCmds[methodName] = method;
        } else if (typeof thisEndpoint[method] === 'function') {
            thisEndpoint.EndpointCmds[methodName] = function (...args) {
                return thisEndpoint[method](...args);
            };
        } else {
            thisEndpoint.log("Cannot add EndpointCmds[" + methodName + "]" + " -> sourceObj[" + method + "] of type " + typeof thisEndpoint[method]);
        }
    }

    /**
     * Send serialized packet data
     * @param {string} drpPacketString JSON string
     * @returns {number} Error code (0 good, 1 wsConn not open, 2 send error)
     */
    SendPacketString(drpPacketString) {
        let thisEndpoint = this;

        if (thisEndpoint.wsConn.readyState !== WebSocket.OPEN)
            //return "wsConn not OPEN";
            return 1;
        try {
            thisEndpoint.wsConn.send(drpPacketString);
            return 0;
        } catch (e) {
            //return e;
            return 2;
        }
    }

    /**
     * 
     * @param {string} serviceName DRP Service Name
     * @param {string} method Service Method
     * @param {Object} params Method Parameters
     * @param {boolean} promisify Should we promisify?
     * @param {function} callback Callback function
     * @param {DRP_RouteOptions} routeOptions Route Options
     * @param {string} serviceInstanceID Execute on specific Service Instance ID
     * @return {Promise} Returned promise
     */
    SendCmd(serviceName, method, params, promisify, callback, routeOptions, serviceInstanceID) {
        let thisEndpoint = this;
        let returnVal = null;
        let token = null;

        if (promisify) {
            // We expect a response, using await; add 'resolve' to queue
            returnVal = new Promise(function (resolve, reject) {
                token = thisEndpoint.AddReplyHandler(function (message) {
                    resolve(message);
                });
            });
        } else if (typeof callback === 'function') {
            // We expect a response, using callback; add callback to queue
            token = thisEndpoint.AddReplyHandler(callback);
        } else {
            // We don't expect a response; leave reply token null
        }
        let packetObj = new DRP_Cmd(serviceName, method, params, token, routeOptions, serviceInstanceID);
        //console.dir(packetObj);
        let packetString = JSON.stringify(packetObj);
        thisEndpoint.SendPacketString(packetString);
        return returnVal;
    }

    /**
     * Send reply to received command
     * @param {string} token Reply token
     * @param {number} status Reply status [0: fail, 1: success, 2: more data coming]
     * @param {any} payload Payload to send
     * @param {DRP_RouteOptions} routeOptions Route options
     * @returns {number} Error string
     */
    SendReply(token, status, payload, routeOptions) {
        let thisEndpoint = this;
        let packetString = null;
        let packetObj = null;

        try {
            packetObj = new DRP_Reply(token, status, payload, routeOptions);
            packetString = JSON.stringify(packetObj);
        } catch (e) {
            packetObj = new DRP_Reply(token, 0, `Failed to stringify response: ${e}`);
            packetString = JSON.stringify(packetObj);
        }
        return thisEndpoint.SendPacketString(packetString);
    }

    /**
     * Process inbound DRP Command
     * @param {DRP_Cmd} cmdPacket DRP Command
     */
    async ProcessCmd(cmdPacket) {
        let thisEndpoint = this;

        var cmdResults = {
            status: 0,
            output: null
        };

        //console.dir(cmdPacket);

        // Make sure params is an Object
        if (!cmdPacket.params || typeof cmdPacket.params !== 'object') cmdPacket.params = {};

        // Override AuthInfo if the remote end is a Consumer
        if (thisEndpoint.AuthInfo && (thisEndpoint.AuthInfo.type === "token" || thisEndpoint.AuthInfo.type === "key")) cmdPacket.params.AuthInfo = thisEndpoint.AuthInfo;

        // Execute method
        try {
            cmdResults.output = await thisEndpoint.DRPNode.ServiceCmd(cmdPacket.serviceName, cmdPacket.method, cmdPacket.params, null, cmdPacket.serviceInstanceID, false, true, thisEndpoint);
            cmdResults.status = 1;
        } catch (err) {
            cmdResults.output = err.message;
        }

        // Reply with results
        if (typeof cmdPacket.token !== "undefined" && cmdPacket.token !== null) {
            let routeOptions = null;
            if (cmdPacket.routeOptions && cmdPacket.routeOptions.tgtNodeID === thisEndpoint.DRPNode.NodeID) {
                routeOptions = new DRP_RouteOptions(thisEndpoint.DRPNode.NodeID, cmdPacket.routeOptions.srcNodeID);
            }
            thisEndpoint.SendReply(cmdPacket.token, cmdResults.status, cmdResults.output, routeOptions);
        }

        //console.dir(cmdResults);
    }

    /**
    * Process inbound DRP Reply
    * @param {DRP_Reply} replyPacket DRP Reply Packet
    */
    async ProcessReply(replyPacket) {
        let thisEndpoint = this;

        //console.dir(replyPacket, { "depth": 10 });

        // Yes - do we have the token?
        if (thisEndpoint.ReplyHandlerQueue.hasOwnProperty(replyPacket.token)) {

            // We have the token - execute the reply callback
            thisEndpoint.ReplyHandlerQueue[replyPacket.token](replyPacket);

            // Delete if the status < 2
            if (!replyPacket.status || replyPacket.status < 2) {
                delete thisEndpoint.ReplyHandlerQueue[replyPacket.token];
            }

        } else {
            // We do not have the token - tell the sender we do not honor this token
        }
    }

    /**
    * Check whether or not to relay the packet
    * @param {DRP_Packet} drpPacket DRP Packet
    * @returns {boolean} Should the packet be relayed?
    */
    ShouldRelay(drpPacket) {
        let thisEndpoint = this;
        /*
         * In order to be relayed, a packet should:
         *   - Have route options
         *   - Specify a tgtNodeID that is not the local Node
         *   - Come from an endpoint that has successfully peered as a Node
         */
        if (drpPacket.routeOptions && drpPacket.routeOptions.tgtNodeID && drpPacket.routeOptions.tgtNodeID !== thisEndpoint.DRPNode.NodeID && thisEndpoint.EndpointType && thisEndpoint.EndpointType === "Node")
            return true;
        else
            return false;
    }

    async ReceiveMessage(rawMessage) {
        let thisEndpoint = this;
        /** @type {DRP_Packet} */
        let drpPacket;
        try {
            drpPacket = JSON.parse(rawMessage);
        } catch (e) {
            thisEndpoint.log(`Received non-JSON message, disconnecting client endpoint[${thisEndpoint.EndpointID}] @ ${thisEndpoint.wsConn._socket.remoteAddress}`);
            thisEndpoint.log(rawMessage);
            thisEndpoint.wsConn.close();
            return;
        }

        // Should we relay the packet?
        if (thisEndpoint.ShouldRelay(drpPacket)) {
            // This is meant for another node
            thisEndpoint.RelayPacket(drpPacket);
            return;
        }

        // Process locally
        switch (drpPacket.type) {
            case 'cmd':
                thisEndpoint.ProcessCmd(drpPacket);
                break;
            case 'reply':
                thisEndpoint.ProcessReply(drpPacket);
                break;
            default:
                thisEndpoint.log("Invalid message.type; here's the JSON data..." + rawMessage);
        }
    }

    /**
     * Relay DRP Packet
     * @param {DRP_Packet} drpPacket Packet to relay
     */
    async RelayPacket(drpPacket) {
        let thisEndpoint = this;
        try {
            // Validate sending endpoint
            if (!thisEndpoint.EndpointID) {
                // Sending endpoint has not authenticated
                throw `sending endpoint has not authenticated`;
            }

            // Validate source node
            if (!thisEndpoint.DRPNode.TopologyTracker.ValidateNodeID(drpPacket.routeOptions.srcNodeID)) {
                // Source NodeID is invalid
                throw `srcNodeID ${drpPacket.routeOptions.srcNodeID} not found`;
            }

            // Validate destination node
            if (!thisEndpoint.DRPNode.TopologyTracker.ValidateNodeID(drpPacket.routeOptions.tgtNodeID)) {
                // Target NodeID is invalid
                throw `tgtNodeID ${drpPacket.routeOptions.tgtNodeID} not found`;
            }

            // Verify whether or not we SHOULD relay the node
            // if (thisEndpoint.DRPNode.IsRegistry() || thisEndpoint.DRPNode.IsRelay())

            let nextHopNodeID = thisEndpoint.DRPNode.TopologyTracker.GetNextHop(drpPacket.routeOptions.tgtNodeID);

            /** @type DRP_Endpoint */
            let targetNodeEndpoint = await thisEndpoint.DRPNode.VerifyNodeConnection(nextHopNodeID);

            // Add this node to the routing history
            drpPacket.routeOptions.routeHistory.push(thisEndpoint.DRPNode.NodeID);

            // We do not need to await the results; any target replies will automatically be routed
            targetNodeEndpoint.SendPacketString(JSON.stringify(drpPacket));
            thisEndpoint.DRPNode.PacketRelayCount++;
            //thisEndpoint.DRPNode.log(`Relaying packet...`);
            //console.dir(drpPacket);

        } catch (ex) {
            // Either could not get connection to node or command send attempt errored out
            thisEndpoint.DRPNode.log(`Could not relay message: ${ex}`);
        }
    }

    GetCmds() {
        return Object.keys(this.EndpointCmds);
    }

    async OpenHandler() {
        if (!this.wsConn) return null;
        this.RemoteAddressPortFamily = `${this.RemoteAddress()}|${this.RemotePort()}|${this.RemoteFamily()}`;
    }

    async CloseHandler() { }

    async ErrorHandler() { }

    Close() {
        if (!this.wsConn) return null;
        this.wsConn.close();
    }

    RemoveSubscriptions() {
        if (!this.wsConn) return null;
        let subscriptionIDList = Object.keys(this.Subscriptions);
        for (let i = 0; i < subscriptionIDList.length; i++) {
            let subscriptionID = subscriptionIDList[i];
            let subscriptionObject = this.Subscriptions[subscriptionID];
            subscriptionObject.Terminate();
            delete this.Subscriptions[subscriptionID];
            this.DRPNode.SubscriptionManager.Subscribers.delete(subscriptionObject);
        }
    }

    IsReady() {
        if (!this.wsConn) return null;
        if (this.wsConn && this.wsConn.readyState === 1)
            return true;
        else
            return false;
    }

    IsConnecting() {
        if (!this.wsConn) return null;
        if (this.wsConn && this.wsConn.readyState === 0)
            return true;
        else
            return false;
    }

    /**
    * @returns {string} Remote Address
    */
    RemoteAddress() {
        if (!this.wsConn) return null;
        let returnVal = null;
        if (this.wsConn && this.wsConn._socket) {
            returnVal = this.wsConn._socket.remoteAddress;
        }
        return returnVal;
    }

    /**
     * @returns {string} Remote Port
     */
    RemotePort() {
        if (!this.wsConn) return null;
        let returnVal = null;
        if (this.wsConn && this.wsConn._socket) {
            returnVal = this.wsConn._socket.remotePort;
        }
        return returnVal;
    }

    /**
     * @returns {string} Remote Family
     */
    RemoteFamily() {
        if (!this.wsConn) return null;
        let returnVal = null;
        if (this.wsConn && this.wsConn._socket) {
            returnVal = this.wsConn._socket.remoteFamily;
        }
        return returnVal;
    }

    /**
     * @returns {number} Uptime in seconds
     */
    UpTime() {
        if (!this.wsConn) return null;
        let currentTime = new Date().getTime();
        return Math.floor((currentTime - this.wsConn.openTime) / 1000);
    }

    /**
     * @returns {number} Ping time in milliseconds
     */
    PingTime() {
        if (!this.wsConn) return null;
        let returnVal = null;
        if (this.wsConn._socket) {
            returnVal = this.wsConn.pingTimeMs;
        }
        return returnVal;
    }

    ConnectionStats() {
        if (!this.wsConn) return null;
        return {
            pingTimeMs: this.PingTime(),
            uptimeSeconds: this.UpTime()
        };
    }

    IsServer() {
        if (!this.wsConn) return null;
        return this.wsConn._isServer;
    }

    log(logMessage) {
        let thisEndpoint = this;
        if (thisEndpoint.DRPNode) {
            thisEndpoint.DRPNode.log(logMessage);
        } else {
            console.log(logMessage);
        }
    }
}

module.exports = DRP_Endpoint;
